bookdown/03-optimization.Rmd

```diff
@@ -231,6 +231,240 @@ Note that we do not expect any differences here compared to the non-tuned approa
+After the initial food is consumed, he checks their performance and kicks the slowest half out of his training regime.
+Then, he increases the available food for the remaining, further trains them until the food is consumed again, only to kick out the worst half once more.
+He repeats this until the one remaining horse gets the rest of the food.
+This means only one horse is fully trained, but on the flip side, he was able to start training with all eight horses.
+On race day, all the horses are put on the starting line.
+But which trainer will have the winning horse?
+The one, who tried to train a maximum amount of horses to their fullest?
+Or the other one, who made assumptions about the training progress of his horses?
+How the training phases may possibly look like is visualized in figure \@ref(fig:03-optimization-012).
+
+```{r 03-optimization-012, eval = TRUE, echo = FALSE, fig.show='hold', out.width = '99%', fig.align = 'center', fig.cap = "Visulization of how the training processes may look like. The left plot corresponds to the non-selective trainer, while the right one to the selective trainer.", warning=FALSE}
+knitr::include_graphics("images/horse_training1.png")
+```
+
+Hyperband works very similar in some ways, but also different in others.
+It is not embodied by one of the trainers in our analogy, but more by the person, who would pay them.
+Hyperband consists of several brackets, each bracket corresponding to a trainer, and we do not care about horses but about hyperparameter configurations of a machine learning algorithm.
+The budget is not in terms of food, but in terms of a hyperparameter of the learner that scales in some way with the computational effort.
+An example is the number of epochs we train a neural network, or the number of iterations in boosting.
+Furthermore, there are not only two brackets (or trainers), but several, each placed at a unique spot between fully explorative of later training stages and extremely selective, equal to higher exploration of early training stages.
+The level of selection aggressiveness is handled by a user-defined parameter called $\eta$.
+So, $1/\eta$ is the fraction of remaining configurations after a bracket removes his worst performing ones, but $\eta$ is also the factor by that the budget is increased for the next stage.
+Because there is a different maximum budget per configuration that makes sense in different scenarios, the user also has to set this as the $R$ parameter.
+No further parameters are required for Hyperband -- the full required budget across all brackets is indirectly given by $$(\lfloor \log_{\eta}{R} \rfloor + 1)^2 * R$$ [@Li2016].
+To give an idea how a full bracket layout might look like for a specific $R$ and $\eta$, a quick overview is given in the following table.
+
+```{r 03-optimization-013, eval = TRUE, echo = FALSE}
+eta = 2
+R = 8
+result = data.frame()
+smax = floor(log(R, eta))
```

Comment:
this has the rounding problem and gives too low smax values for some values of eta and R. this is already fixed in the hyperband code but maybe you want to fix it here as well.

Reply:
I don't think it's really necessary in this place, since the rounding error only happens if R is a quotient (for example `floor(log(81, 3))` has no rounding error, but `floor(log(8.1/0.1, 3))` has).
Even if i wanted to fix it here, it would not be possible in a "nice" way, since the fix for the latter would be `floor(log(8.1, 3) - log(0.1, 3))` and I can only write it like that if the quotient is known

Reply:
```r
> log(243, 3)
[1] 5
> floor(log(243, 3))
[1] 4
```

Reply:
I was about to push a fix when I realized that this code is not meant to be production code but only meant to showcase the hyperband formula for these very specific values. By adding a fix for more arbitrary values, we may shift the user's attention to something he should not care about or even confuse her/him if s/he is unfamiliar with machine precision.